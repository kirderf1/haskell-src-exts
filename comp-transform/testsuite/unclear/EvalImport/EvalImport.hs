{-# LANGUAGE ComposableTypes #-}

module Main where

--TODO Add Common file to test when we've got module imports figured out
import Common

eval -: Expr -> Int


ext eval for Value where 
    eval (Const i) = i
   
ext eval for Op where
    eval (Add e1 e2) = eval e1 + eval e2
    eval (Mult e1 e2) = eval e1 * eval e2
