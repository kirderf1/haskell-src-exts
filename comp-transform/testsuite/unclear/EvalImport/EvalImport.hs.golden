{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FlexibleContexts #-}
module Main where
import qualified Data.Comp
import qualified Data.Comp.Derive
import qualified Data.Comp.Show ()
import qualified Data.Comp.Equality ()
import Common

class Eval f where
        eval' :: (Eval g) => Expr -> Int

eval :: (Eval g) => Expr -> Int
eval = eval' . Data.Comp.unTerm

{-# NOINLINE eval #-}

$( Data.Comp.Derive.derive [Data.Comp.Derive.liftSum] [''Eval] )

instance Eval Value where
        eval' (Const i) = i

instance Eval Op where
        eval' (Add e1 e2) = eval e1 + eval e2
        eval' (Mult e1 e2) = eval e1 * eval e2
